File: app-flutter/lib/core/services/continue_watching_service.dart
import 'dart:convert';

import 'package:nebras_mobile_app/features/content/model/content_model.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// تقدّم غير مكتمل — **محليّ فقط** (لا يُرفع إلى Firebase).
///
/// يُغذّي رفّ «تابع التصفّح» في الصفحة الرئيسيّة.
/// يُخفى العنصر إذا اكتمل (>90%) أو مرّ عليه >14 يوماً.
class ContinueWatchingService {
  ContinueWatchingService._();

  static final ContinueWatchingService instance = ContinueWatchingService._();

  static const _kEntries = 'continue_watching_v1';
  static const _maxEntries = 24;
  static const _staleAfter = Duration(days: 14);

  SharedPreferences? _prefs;

  Future<void> init() async {
    _prefs ??= await SharedPreferences.getInstance();
  }

  Future<void> saveProgress({
    required String contentId,
    required int positionMs,
    required int durationMs,
    required String title,
    required String thumbnailUrl,
    required ContentType type,
  }) async {
    if (contentId.trim().isEmpty || durationMs <= 0) return;
    await init();
    final ratio = positionMs / durationMs;
    if (ratio >= 0.9) {
      await remove(contentId);
      return;
    }
    if (ratio < 0.02) return;

    final map = await _readMap();
    map[contentId] = {
      'contentId': contentId,
      'positionMs': positionMs,
      'durationMs': durationMs,
      'title': title,
      'thumbnailUrl': thumbnailUrl,
      'type': type.name,
      'savedAt': DateTime.now().toIso8601String(),
    };
    final sorted = map.entries.toList()
      ..sort((a, b) {
        final at = DateTime.tryParse(
              '${a.value['savedAt']}',
            ) ??
            DateTime.fromMillisecondsSinceEpoch(0);
        final bt = DateTime.tryParse(
              '${b.value['savedAt']}',
            ) ??
            DateTime.fromMillisecondsSinceEpoch(0);
        return bt.compareTo(at);
      });
    final trimmed = <String, dynamic>{
      for (final e in sorted.take(_maxEntries)) e.key: e.value,
    };
    await _prefs!.setString(_kEntries, jsonEncode(trimmed));
  }

  Future<void> remove(String contentId) async {
    await init();
    final map = await _readMap();
    if (map.remove(contentId) != null) {
      await _prefs!.setString(_kEntries, jsonEncode(map));
    }
  }

  /// مسح كامل — تسجيل الخروج أو «نسيان اهتماماتي».
  Future<void> clearAll() async {
    await init();
    await _prefs!.remove(_kEntries);
  }

  Future<List<ContinueWatchingEntry>> activeEntries() async {
    await init();
    final map = await _readMap();
    final now = DateTime.now();
    final out = <ContinueWatchingEntry>[];
    final staleKeys = <String>[];

    map.forEach((id, raw) {
      final savedAt = DateTime.tryParse('${raw['savedAt']}');
      if (savedAt == null ||
          now.difference(savedAt) > _staleAfter) {
        staleKeys.add(id);
        return;
      }
      final durationMs = (raw['durationMs'] as num?)?.toInt() ?? 0;
      final positionMs = (raw['positionMs'] as num?)?.toInt() ?? 0;
      if (durationMs <= 0) {
        staleKeys.add(id);
        return;
      }
      final ratio = positionMs / durationMs;
      if (ratio >= 0.9) {
        staleKeys.add(id);
        return;
      }
      out.add(
        ContinueWatchingEntry(
          contentId: id,
          title: '${raw['title'] ?? ''}',
          thumbnailUrl: '${raw['thumbnailUrl'] ?? ''}',
          type: _parseType('${raw['type']}'),
          positionMs: positionMs,
          durationMs: durationMs,
          savedAt: savedAt,
        ),
      );
    });

    for (final k in staleKeys) {
      map.remove(k);
    }
    if (staleKeys.isNotEmpty) {
      await _prefs!.setString(_kEntries, jsonEncode(map));
    }

    out.sort((a, b) => b.savedAt.compareTo(a.savedAt));
    return out;
  }

  Future<Map<String, dynamic>> _readMap() async {
    final raw = _prefs?.getString(_kEntries);
    if (raw == null || raw.isEmpty) return {};
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map) return {};
      return decoded.map(
        (k, v) => MapEntry(k.toString(), Map<String, dynamic>.from(v as Map)),
      );
    } catch (_) {
      return {};
    }
  }

  ContentType _parseType(String name) {
    for (final t in ContentType.values) {
      if (t.name == name) return t;
    }
    return ContentType.video;
  }
}

class ContinueWatchingEntry {
  final String contentId;
  final String title;
  final String thumbnailUrl;
  final ContentType type;
  final int positionMs;
  final int durationMs;
  final DateTime savedAt;

  const ContinueWatchingEntry({
    required this.contentId,
    required this.title,
    required this.thumbnailUrl,
    required this.type,
    required this.positionMs,
    required this.durationMs,
    required this.savedAt,
  });

  double get progressRatio =>
      durationMs > 0 ? (positionMs / durationMs).clamp(0.0, 1.0) : 0.0;
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/service/ContinueWatchingService.kt
package com.nebras.mobile.core.service

import com.nebras.mobile.core.data.LocalStore
import com.nebras.mobile.core.model.ContentType
import org.json.JSONObject
import kotlin.math.min

/**
 * تقدّم غير مكتمل — **محليّ فقط** (لا يُرفع إلى Firebase).
 *
 * يُغذّي رفّ «تابع التصفّح» في الصفحة الرئيسيّة.
 * يُخفى العنصر إذا اكتمل (>90%) أو مرّ عليه >14 يوماً.
 */
object ContinueWatchingService {

    private const val KEY_ENTRIES = "continue_watching_v1"
    private const val MAX_ENTRIES = 24
    private const val STALE_AFTER_MILLIS = 14L * 24 * 60 * 60 * 1000

    private lateinit var store: LocalStore

    fun init(localStore: LocalStore) {
        store = localStore
    }

    fun saveProgress(
        contentId: String,
        positionMs: Long,
        durationMs: Long,
        title: String,
        thumbnailUrl: String,
        type: ContentType,
    ) {
        if (contentId.trim().isEmpty() || durationMs <= 0) return
        val ratio = positionMs.toDouble() / durationMs
        if (ratio >= 0.9) {
            remove(contentId)
            return
        }
        if (ratio < 0.02) return

        val map = readMap()
        map.put(
            contentId,
            JSONObject().apply {
                put("contentId", contentId)
                put("positionMs", positionMs)
                put("durationMs", durationMs)
                put("title", title)
                put("thumbnailUrl", thumbnailUrl)
                put("type", type.name)
                put("savedAt", System.currentTimeMillis())
            },
        )

        // نُبقي الأحدث حفظاً فقط ضمن السقف.
        val sorted = map.keys().asSequence().toList()
            .mapNotNull { key -> map.optJSONObject(key)?.let { key to it } }
            .sortedByDescending { it.second.optLong("savedAt", 0L) }
            .take(MAX_ENTRIES)

        val trimmed = JSONObject()
        sorted.forEach { (key, value) -> trimmed.put(key, value) }
        store.putJsonObject(KEY_ENTRIES, trimmed)
    }

    fun remove(contentId: String) {
        val map = readMap()
        if (map.has(contentId)) {
            map.remove(contentId)
            store.putJsonObject(KEY_ENTRIES, map)
        }
    }

    /** مسح كامل — تسجيل الخروج أو «نسيان اهتماماتي». */
    fun clearAll() = store.remove(KEY_ENTRIES)

    fun activeEntries(): List<ContinueWatchingEntry> {
        val map = readMap()
        val now = System.currentTimeMillis()
        val out = mutableListOf<ContinueWatchingEntry>()
        val staleKeys = mutableListOf<String>()

        for (id in map.keys().asSequence().toList()) {
            val raw = map.optJSONObject(id) ?: run { staleKeys.add(id); continue }
            val savedAt = raw.optLong("savedAt", 0L)
            if (savedAt <= 0L || now - savedAt > STALE_AFTER_MILLIS) {
                staleKeys.add(id)
                continue
            }
            val durationMs = raw.optLong("durationMs", 0L)
            val positionMs = raw.optLong("positionMs", 0L)
            if (durationMs <= 0) {
                staleKeys.add(id)
                continue
            }
            if (positionMs.toDouble() / durationMs >= 0.9) {
                staleKeys.add(id)
                continue
            }
            out.add(
                ContinueWatchingEntry(
                    contentId = id,
                    title = raw.optString("title"),
                    thumbnailUrl = raw.optString("thumbnailUrl"),
                    type = parseType(raw.optString("type")),
                    positionMs = positionMs,
                    durationMs = durationMs,
                    savedAt = savedAt,
                ),
            )
        }

        if (staleKeys.isNotEmpty()) {
            staleKeys.forEach(map::remove)
            store.putJsonObject(KEY_ENTRIES, map)
        }

        return out.sortedByDescending { it.savedAt }
    }

    private fun readMap(): JSONObject = store.getJsonObject(KEY_ENTRIES) ?: JSONObject()

    private fun parseType(name: String): ContentType =
        ContentType.entries.firstOrNull { it.name.equals(name, ignoreCase = true) }
            ?: ContentType.VIDEO
}

data class ContinueWatchingEntry(
    val contentId: String,
    val title: String,
    val thumbnailUrl: String,
    val type: ContentType,
    val positionMs: Long,
    val durationMs: Long,
    val savedAt: Long,
) {
    val progressRatio: Float
        get() = if (durationMs > 0) {
            min(1.0, positionMs.toDouble() / durationMs).toFloat().coerceAtLeast(0f)
        } else {
            0f
        }
}
